### PR TITLE
Fixing potential deadlock during map loading

### PR DESCRIPTION
**Pull request by ayrton04**
FYI @jasonimercer @p6chen @skaynama

### CODE REVIEW

File: costmap_2d/plugins/static_layer.cpp [hunk 1]
```diff
@@ -174,10 +174,18 @@ void StaticLayer::incomingMap(const nav_msgs::OccupancyGridConstPtr& new_map)
       master->getOriginY() != new_map->info.origin.position.y ||
       !layered_costmap_->isSizeLocked()))
   {
+    // Calling resizeMap on the layered_costmap_ is going to update the global/master map, not the local one for this
+    // layer. It also will lock the mutex around the global/master map, so we should unlock the local one first to
+    // avoid deadlock situations.
+    lock.unlock();
```
Comment (jasonimercer) [inline costmap_2d/plugins/static_layer.cpp]: if we've unlocked the mutex here can we be sure that size_x and size_y are safe to use down on lines 204 and 206?

Comment (ayrton04) [inline costmap_2d/plugins/static_layer.cpp]: When we call layered_costmap_->resizeMap, it goes up to the layered cost map, loops through the layers (including this one), and calls matchSize. So if something changed the local costmaps size between the unlock statement on line 180 and the resize call on 184, we'd be fine, as the size will immediately be set back to the size that we passed to resizemap. However, if someone managed to somehow resize the map in this layer (though I don't know if there's a current sequence of events that could lead to this) between line 184 and line 188, then it is conceivable that the local costmap could be resized, yes.

As a workaround, we can simply remove the else in the statement on 190. If the call on 184 works and there is no resize in between there and 188, then the clause will fail, as matchSize will make everything line up. If something does manage to change the size, then we'd enter that block and fix it on line 197. What do you think?

Comment (jasonimercer) [inline costmap_2d/plugins/static_layer.cpp]: I like it, that sounds like a good fix.

**Comment by skaynama** [comment]
LGTM.

**Pull request merged**